src/components/PopupBase/PopupBase.js

```diff
@@ -0,0 +1,172 @@
+    }
+
+    componentDidMount() {
+        document.addEventListener('click', this.handleWindowClick);
+        document.addEventListener('keydown', this.handleKeyPress, false);
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        const { isOpened } = this.state;
+        if (isOpened && !prevState.isOpened) {
+            this.createPopperInstance();
+        }
+        if (!isOpened && prevState.isOpened) {
+            this.destroyPopperIstance();
+        }
+    }
+
+    componentWillUnmount() {
+        document.removeEventListener('click', this.handleWindowClick);
+        document.removeEventListener('keydown', this.handleKeyPress, false);
+    }
+
+    setPopupVisibility(shouldBeOpen) {
+        const { isOpened } = this.state;
+        if (shouldBeOpen !== isOpened) {
+            this.setState({ isOpened: shouldBeOpen });
+        }
+    }
+
+    getArgs() {
+        const { isOpened } = this.state;
+        return {
+            setPopupVisibility: this.setPopupVisibility,
+            isOpened,
+        };
+    }
+
+    handleWindowClick(event) {
+        if (!this.popupRef.current) return;
+        const wasPopupClicked = PopupBase.checkDOMPath(
+            event.target,
+            elem => elem.dataset && !!elem.dataset.popup
+        );
+        const wasOutsideClicked = !this.popupRef.current.contains(event.target);
+
+        const wasAnchorClicked =
+            this.anchorRef.current && this.anchorRef.current.contains(event.target);
+        if (!wasPopupClicked && wasOutsideClicked && !wasAnchorClicked) {
+            this.close();
+        }
+    }
+
+    handleKeyPress(event) {
+        if (event.key === ESCAPE_KEY) {
+            this.close();
+        }
+    }
+
+    close() {
+        const { isOpened } = this.state;
+        if (isOpened === true) {
+            this.setState({ isOpened: false });
+        }
+    }
+
+    createPopperInstance() {
+        if (!this.anchorRef.current || !this.popupRef.current) return;
+        const { placement } = this.props;
+        this.destroyPopperIstance();
+        this.popper = new PopperJS(this.anchorRef.current, this.popupRef.current, {
+            placement,
+        });
+    }
+
+    destroyPopperIstance() {
+        if (!this.popper) return;
+        this.popper.destroy();
+        this.popper = undefined;
+    }
+
+    renderAnchor() {
+        const { anchorRenderer } = this.props;
+        const anchorElem = anchorRenderer(this.getArgs());
+        return React.cloneElement(anchorElem, { ref: this.anchorRef });
+    }
+
+    renderPopup() {
+        const { isOpened } = this.state;
+        if (!isOpened) return null;
+        const { popupRenderer } = this.props;
+        const popupElem = popupRenderer(this.getArgs());
+        const popupElemWithProps = React.cloneElement(popupElem, {
+            ref: this.popupRef,
+            'data-popup': 'true',
+        });
+        return ReactDOM.createPortal(popupElemWithProps, document.body);
+    }
```

Comment:
if (isOpened) {
        const { popupRenderer } = this.props;
        const popupElem = popupRenderer(this.getArgs());
        const popupElemWithProps = React.cloneElement(popupElem, {
            ref: this.popupRef,
            'data-popup': 'true',
        });

        return ReactDOM.createPortal(popupElemWithProps, document.body);
}

return null;